controllers/appmesh/cloudmap_controller.go

```diff
@@ -0,0 +1,888 @@
+		cloudMapInstanceCache: cache.NewTTLStore(func(obj interface{}) (string, error) {
+			return obj.(*cloudMapInstanceCacheItem).key, nil
+		}, 300*time.Second),
+		nameSpaceIDCache: cache.NewTTLStore(func(obj interface{}) (string, error) {
+			return obj.(*cloudmapNamespaceCacheItem).key, nil
+		}, 60*time.Second),
+		serviceIDCache: cache.NewTTLStore(func(obj interface{}) (string, error) {
+			return obj.(*cloudmapServiceCacheItem).key, nil
+		}, 60*time.Second),
+		enqueueRequestsForPodEvents: virtualnode.NewEnqueueRequestsForPodEvents(k8sClient, log),
+	}
+}
+
+// +kubebuilder:rbac:groups=appmesh.k8s.aws,resources=virtualnodes,verbs=get;list;watch
+// +kubebuilder:rbac:groups=appmesh.k8s.aws,resources=virtualnodes/status,verbs=get
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
+
+func (r *cloudMapReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	return runtime.HandleReconcileError(r.reconcile(req), r.log)
+}
+
+func (r *cloudMapReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appmesh.VirtualNode{}).
+		Watches(&source.Kind{Type: &corev1.Pod{}}, r.enqueueRequestsForPodEvents).
+		Complete(r)
+}
+
+func (r *cloudMapReconciler) reconcile(req ctrl.Request) error {
+	ctx := context.Background()
+	log := r.log.WithValues("Virtualnode", req.NamespacedName)
+
+	vNode := &appmesh.VirtualNode{}
+	if err := r.k8sClient.Get(ctx, req.NamespacedName, vNode); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	log.V(2).Info("VNode: ", "awsName: ", vNode.Spec.AWSName,
+		"vNode.PodSelector: ", vNode.Spec.PodSelector)
+	if !vNode.DeletionTimestamp.IsZero() {
+		return r.deleteCloudMapResources(ctx, vNode)
+	}
+	return r.reconcileVirtualNodeWithCloudMap(ctx, vNode)
+}
+
+func (r *cloudMapReconciler) reconcileVirtualNodeWithCloudMap(ctx context.Context, vNode *appmesh.VirtualNode) error {
+
+	if !r.isCloudMapEnabledForVirtualNode(vNode) {
+		return errors.Errorf("CloudMap not enabled for virtualNode: %s", vNode.Name)
+	}
+
+	if !r.isPodSelectorDefinedForVirtualNode(vNode) {
+		return errors.Errorf("PodSelector not defined for virtualNode: %s", vNode.Name)
+	}
+
+	cloudmapConfig := vNode.Spec.ServiceDiscovery.AWSCloudMap
+	creatorRequestID := vNode.UID
+
+	var podsList corev1.PodList
+	var listOptions client.ListOptions
+	listOptions.LabelSelector, _ = metav1.LabelSelectorAsSelector(vNode.Spec.PodSelector)
+	listOptions.Namespace = vNode.Namespace
+
+	if err := r.k8sClient.List(ctx, &podsList, &listOptions); err != nil {
+		r.log.Error(err, "Couldn't retrieve pods for VirtualNode")
+		return err
+	}
+
+	r.log.V(2).Info("Pods found matching this VNode: ", "Total Pods: ", len(podsList.Items))
+	if err := r.finalizerManager.AddFinalizers(ctx, vNode, k8s.FinalizerAWSCloudMapResources); err != nil {
+		r.log.Error(err, "..while adding cloudmap resource finalizer")
+		return err
+	}
+
+	_, err := r.createCloudMapService(ctx, cloudmapConfig, string(creatorRequestID))
```

Comment:
better to be named as "reconcileCloudMapService" or "getOrCreateCloudMapService)

Reply:
renamed it